charm/src/actions.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""HTTPRequest LEGO provider charm actions."""
+
+# pylint: disable=protected-access
+
+import logging
+import secrets
+
+import ops
+import xiilib.django
+
+logger = logging.getLogger(__name__)
+
+
+class NotReadyError(Exception):
+    """Exception thrown when needed resources are not ready."""
+
+
+class Observer(ops.Object):
+    """Charm actions observer."""
+
+    def __init__(self, charm: xiilib.django.Charm):
+        """Initialize the observer and register actions handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "actions-observer")
+        self.charm = charm
+
+        charm.framework.observe(charm.on.create_user_action, self._create_or_update_user)
+        charm.framework.observe(charm.on.allow_domains_action, self._allow_domains)
+        charm.framework.observe(charm.on.revoke_domains_action, self._revoke_domains)
+        charm.framework.observe(charm.on.list_domains_action, self._list_domains)
+
+    def _generate_password(self) -> str:
+        """Generate a new password.
+
+        Returns: the new password.
+        """
+        return secrets.token_urlsafe(30)
+
+    def _execute_command(self, command: list[str], event: ops.ActionEvent) -> None:
+        """Prepare the scripts for exxecution.
+
+        Args:
+            command: the management command to execute.
+            event: the event triggering the original action.
+
+        Raises:
+            ExecError: if an error occurs while executing the script
+        """
+        container = self.charm.unit.get_container(self.charm._CONTAINER_NAME)
```

Comment:
@weiiwang01 `_CONTAINER_NAME` seems like it is reasonable to make this a public attribute? I also thought about whether we should provide a easier way to get the container, since any customisation would need that as well, I'm slightly hesitant because the current approach would be more like a normal charm developer, though at the same time I do think we should have nice interfaces for common actions. Could we please create a story to create a spec for this specific interface?

Reply:
Yes, the problem is that we do not yet have a set of stable APIs to expose to users. Even event handler methods like `_on_config_changed` are technically private. So, we definitely need one.

Reply:
The dev experience was pretty good, so not much tbh. I think you shuold consider making `self.charm._BASE_DIR` public too, and probably provide a constant or something to retrieve the user and group the service is run with